miner/worker.go

```diff
@@ -1134,6 +1134,9 @@ func (w *worker) commit(env *environment, interval func(), update bool, start ti
 		if interval != nil {
 			interval()
 		}
+		// Create a local environment copy, avoid the data race with snapshot state.
+		// https://github.com/ethereum/go-ethereum/issues/24299
+		env := env.copy()
```

Comment:
A few lines up, at line `1118`, we already do `w.commit(work.copy(), w.fullTaskHook, true, start)`. Are you sure this fixes it?

Reply:
Yup, in line `1118`, we create the copy for commit function to isolate the mutation between the `commit` function and other logic in miner.

However, in the `commit` function, there are two places can read/write the passed env in the same time. One is `updateSnapshot` which inits the pending state snapshot. Another is `resultLoop` for mutating the state root.

Reply:
I still don't see how this fixes it. Don't the two still use the same instance, only this time a copy of the one that was passed in?

Reply:
If mining is enabled, then the *copy of passed env* is passed to the sealer and the passed env is passed to the pending state.